### ISSUE TITLE
Date.prototype.toJSON has incorrect return type

### ISSUE THREAD

**Issue by siefkenj** [microsoft/TypeScript#43639]
# Bug Report

`Date.prototype.toJSON()` can return a `string` or `null` according to the spec https://tc39.es/ecma262/#sec-date.prototype.tolocaledatestring. For example `new Date("").toJSON()` returns `null`.

Currently, the return type of `toJSON()` is specified as a string:
https://github.com/microsoft/TypeScript/blob/131875bb849c0a9c56c26d09881453aace0cbfed/lib/lib.es5.d.ts#L879

### 🕗 Version & Regression Information

Typescript 4.2.3

### 💻 Code

```ts
new Date("").toJSON().slice(0,5);
```
produces an error at runtime since `new Date("").toJSON()` is `null`, but doesn't produce a typescript error because `new Date("").toJSON()` is assumed to be string.

### 🙂 Expected behavior

TypeScript should warn about the potentially `null` value when an arbitrary string is passed into `Date`. However, if the `Date` object can be guaranteed to be valid (for example `new Date()`), the return type of `toJSON()` should be `string`.

**Comment by andrewbranch** [comment]
You’re right, but making these kinds of changes often breaks a lot of people for not a whole lot of payoff. (People may have other validation systems in place that ensures their Dates are valid, which would make it a huge pain for them if all of a sudden they had to pretend that all their Dates might _not_ be valid.) This is the sort of thing we want to hear a lot of demand for first.

**Comment by RA80533** [comment]
Is there a supported way for overriding the built-in types that have this sort of issue?

**Comment by andrewbranch** [comment]
Unfortunately, there’s not a super clean, easy way to make a built-in type _wider_ than it’s already declared 😕. This sort of thing makes me want to revisit the idea of having “strict” variants of lib files that people could opt into with the `lib` compiler option, so we could make these kind of changes without breaking the world.

I guess the easiest option I would recommend is just writing a wrapper function around built-ins whose types are too loose, and always use your wrapper functions instead of the built-ins (you could enforce this with an ESLint rule):

```ts
export function dateToJSON(date: Date): string | null {
  return date.toJSON();
}
```

**Comment by siefkenj** [comment]
I think a `--strictLibs` style option would be a good compromise. Is there currently a list of non-spec-compliant types that are not being corrected due to compatibility claims?

**Comment by andrewbranch** [comment]
There’s no list, but the biggest category is that every `any` in a covariant position would become `unknown`. I think the barrier to doing it now is really just a matter of figuring out how to automate everything so we aren’t actually maintaining 2x the number of lib files. In particular, the DOM libs are generated from IDL files, so inserting a bunch of manual editing into that process is kind of a non-starter. And changing `unknown` to `any` or vice versa is easy, but this `string | null` return type is a good example of something we’d have to think about. Not _every_ nullable type should be made non-null in the non-strict libs. There are a lot of open questions about how the whole process should work.

**Comment by Joozty** [comment]
Would be really nice to have correct return types in `toJSON()` and `JSON.strigify()`. Today we run into a really bad bug on production because of this. I would expect that TS warned us. Sadly that was not the case. 😢

**Comment by sylann** [comment]
Maybe this was obvious for all here, but it might not be for all readers so I feel compelled to indicate that ‘Date.toJSON()’ does not just return null for the empty string. It does so for any ‘Invalid Date’, which I cannot imagine how nobody would have encountered this case at some point.

I hear your concerns but I don’t think it’s ok to not fix a bug because **maybe** people’s code is correct and the fix could annoy them.

I feel like this is more likely that many have a bug but don’t know it yet, or the code is already loosely typed and this wouldn’t affect them, or they already noticed and wrapped the code like you proposed.

I encountered this yesterday and immediately checked the spec, so it was not a big deal. But if it did not happen early enough it could have been worse. Will never know.
Just my 2 cents.
